Adding/expanding hold_mode mechanism to generic_thermostat

alexbarcelo opened the pull request:
**Description:**

This PR adds support for several `hold` modes (currently: `away`, `home` [default], `vacation` and `sleep`). The mode can be set through the recommended mechanism `set_hold_mode`. 

Each `hold_mode` has its own temperature tracking (being the `home` the default one, also called the "target temperature").

Before doing the PR in the documentation I would like to address any issues with the behaviour or the semantics.

**Example entry for `configuration.yaml`:**
```yaml
# Example configuration.yaml entry
climate:
  - platform: generic_thermostat
    name: Study
    heater: switch.study_heater
    target_sensor: sensor.study_temperature
    target_temp: 21.5
    hold_temps:
      away: 16.5
      vacation: 6
```

**Checklist:**

  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)
  - [x] Local tests with `tox` run successfully.

alexbarcelo committed:
Adding 'away' hold mechanism to generic_thermostat

alexbarcelo committed:
Adding simple tests for the 'away' mechanism

alexbarcelo committed:
added 'arrive home' scenario to hold mode tests

Danielhiversen commented:
Thanks.
I think it should be possible to change the away temp with a service call.

alexbarcelo commented:
@Danielhiversen I thought about that, but could not decide exactly how (both technical and logistics). Some other climate components (at least, ecobee) do something about "away temperatures", but I don't know how they do that (I checked the source code, but I am not sure how it really works).

Do you think that it is worth adding the general service call to the climate component? I struggle to understand what is an "operation mode" and what is a "hold mode", in order to decide what should go into the `climate.set_temperature` service call. And whether all hold modes should (potentially) have a temperature or not.

alexbarcelo committed:
proposal of hold_mode management on climate platform

alexbarcelo committed:
modifying and generalizing the generic_thermostat hold_mode management

alexbarcelo committed:
fixing tests and adding some configuration extra ones

MartinHjelmare reviewed (commented):
Set temperature service schema needs to be updated also. There's a caveat to the current schema that temperature is not a required attribute, where at least one of the temperatures should be required.

Danielhiversen commented:
I don't see much benefit of adding this fynctionality. It adds a lot of complexity to the code, so we should  be sure it will be useful. 

I am thinking it easily can be done by an automation? With this pr you would still need a script to change the different temperatures and mode.

alexbarcelo commented on homeassistant/components/climate/generic_thermostat.py:
```diff
@@ -129,17 +155,70 @@ def current_operation(self):
+            self._target_temp = temperature
+
+        # Store the temperature in the hold_temps dictionary
+        self._hold_temps[hold_mode] = temperature
```
You don't set a hold mode through the `set_temperature` service call. You do so by using the `set_hold_mode`, don't you?

MartinHjelmare replied:
But you've added hold mode to the set temperature service? 

Temperature used to be required but the current service schema doesn't enforce that properly. That's something we should fix, albeit not in this PR. Still you need to currently consider catering for that problem.

At minimum you should update the service schema with the the new optional attribute.

alexbarcelo replied:
> But you've added hold mode to the set temperature service?

Yes, because you can set a temperature for a certain hold mode. But you do not change the hold mode from the `set_temperature` service call. You change the hold_mode through `set_hold_mode`, and set the temperature for hold modes through `set_temperature`.

> At minimum you should update the service schema with the the new optional attribute.

Yes, that is true. Completely missed that.

MartinHjelmare replied:
I'm only talking about the set temperature service call attributes, ie what hold mode should be connected to what temperature. Have I understood it's use correctly? I know that you catch the case of a missing temperature for a hold mode when selecting that hold mode. But I think it would be good to also catch and inform the user when trying to set a hold's temperature without temperature, although it would be weird to try and do that without a temperature. But never underestimate the user's imagination. :smile:

alexbarcelo replied:
Sorry! Now I understand. Yes, I did the catch because it was in the original thermostat code. I wasn't sure why there was a `None` catch, decided to leave it there. Not sure if makes sense or it would be better to log a ERROR and completely ignore the `set_temperature` call without temperature.

MartinHjelmare replied:
Log error and ignore is good, until we fix this in the service schema properly. You should still also add the hold mode attribute as an optional attribute in the service schema.

alexbarcelo commented:
You can do the same by using automations and a lot of sliders, yes you can. In fact, you could completely throw away the `generic_thermostat` and use automations. But the features were there, in the climate component, and they were used by some other climate platforms.

I know that the balance point is somewhat difficult to agree on. My initial use case was: I want to be able to set the temperature through the `generic_thermostat` and I want to have an "away" mode in order to shut down. But the "home" mode should restore the temperature that originally was there. To achieve this with automations I would need an intermediate variable to store the temperature (or use some mechanism that I could not find) and the automation should restore that. If there is no easier way, then I would drop the `generic_thermostat` entirely and do it by hand. Which may be the best solution to may problem, but I was thinking that maybe I am not the only one and `generic_thermostat` may benefit from that.

alexbarcelo committed:
fixing service call schema and explicit (simpler) required temp attribute

pvizeli commented:
I see a lot of climate platform they make he own fake state stuff. My question, If that possible to add this to base if the device not support a real mode for that? @balloob, @turbokongen ?

alexbarcelo commented:
Given that maybe there are some structural challenges on that, and different point of views, I started this topic:
https://community.home-assistant.io/t/balancing-documentation-and-service-calls-across-climate-component/11837

I like the idea of adding state stuff to platforms (as can be seen on my PR). That is beyond my initial idea, but it suits me. I will keep an eye for comments on this PR, and possible improvements. Or redo it completely. Or throw it away and keep it as a `custom_component` in my setup :)

andrey-git commented:
Any updates on this PR?

alexbarcelo commented:
I was waiting on some feedback regarding the direction of the PR. I can revert to the first away-only simple mechanism (addressing the concerns of @Danielhiversen on code complexity), or I can expand a little bit the selection of hold modes (if the more complex behaviour of hold modes is worth it, in the `generic_thermostat` platform context).

I have no problem on changing / readapting the PR. Right now, _It Works For Me_ :tm:, but there is room for improvement before a merge.

(Sorry if I am being overzealous, but I hate spending time trying to raise code quality and end up throwing it away because nobody benefits from it).

balloob commented:
@Danielhiversen could you give some feedback regarding the direction of this PR?

alexbarcelo commented:
I just have found #6258, it seems that the ecobee has changed a little bit in version 0.40 and maybe some ideas that I wanted to introduce in the generic thermostat are now more "standard" in the ecobee implementation.

@Duoxilian, may I ask your opinion on this?

Following the lead of #6258 would be a good idea, IMHO. Maybe some specific platform variables and constants can be moved into the component.

Duoxilian commented:
Alex,

I am sorry, I am not quite sure what you are asking my opinion about. I
implemented #6258, so I have a natural bias for that one and recommend that
route ;-)

Please advise what ideas you want to introduce and I can give my 2c from an
ecobee point of view...

Cheers...

On Fri, Mar 17, 2017 at 2:14 AM, Alex Barcelo <notifications@github.com>
wrote:

> I just have found #6258
> <https://github.com/home-assistant/home-assistant/pull/6258>, it seems
> that the ecobee has changed a little bit in version 0.40 and maybe some
> ideas that I wanted to introduce in the generic thermostat are now more
> "standard" in the ecobee implementation.
>
> @Duoxilian <https://github.com/Duoxilian>, may I ask your opinion on this?
>
> Following the lead of #6258
> <https://github.com/home-assistant/home-assistant/pull/6258> would be a
> good idea, IMHO. Maybe some specific platform variables and constants can
> be moved into the component.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/5832#issuecomment-287283720>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AXZhlhTfsHDwruzVft0W5E6uGIfJa2z0ks5rmjK-gaJpZM4L8Taq>
> .
>

Danielhiversen commented:
It makes the code more complicated, but we have good tests for the component and this new functionality is well tested. 
I am doing something similar with different automations to day, but I will start using this functionality instead.
So I think it is worth adding it. :+1:

alexbarcelo commented:
@Duoxilian, some notes/questions/stuff from the top of my head:

Does #6258 approach make sense without ecobee specific hardware? What is the "black magic" done by ecobee and what is the link between the hass platform and the physical device? The idea of using more hold modes is something that would make sense (like a sleep hold mode)? What should generic climate implement to be somewhat compatible with the ecobee (scenario where a user migrates from one platform to another because they change physical device)? Are there implementation decisions worth doing with that in mind? Should I ignore completely the #6258 and do something "sensible" in the generic_climate? (I would hate to go that path, because maintenance and standardization would suck, IMHO, but maybe I am wrong and I am looking into the wrong direction).

Note that I have no idea of what ecobee does (more specifically, how it does it), but I feel that ecobee climate component is pretty much alive and I thought that it would be an organic way to approach the generic_climate. Again, I may be completely wrong on that approach.

Duoxilian commented:
@alexbarcelo: Let me give some background on ecobee, and you can then decide where you want to take this conversation. I assume that most modern thermostats provide something similar, but of course the devil is in the details. Please also review the documentation on home-assistant.io (https://home-assistant.io/components/climate.ecobee/ and click on the Ecobee Thermostat link). I tried to outline some key concepts there. Please look at the definitions of _target temperature_, _climate_, _hold_, and _operation mode_.

I assume that similar concepts exist for other smart thermostats. I would start with a basic definitions of these concepts. If one knew a number of existing thermostats, one could see what a common definition might look like and what variations there are.

Let's look at the example of _hold_, as this was subject of several recent updates I did for the ecobee climate device. Regarding a hold, you might want to ask yourself the following questions:
- How is the target temperature defined (e.g., based on a reference climate, explicitly set, etc.)?
- Are there other thermostat properties which are set in hold mode (e.g., fan interval)?
- How long does the hold last (e.g., permanently, until next thermostat transition, until a specific time)?

andrey-git commented:
@alexbarcelo Alex, any updates on this PR?

pvizeli commented:
Leave a comment
